Fix binary compatibility issue

andreas-bok-sociomantic opened the pull request:
Avoid breaking breaking the binary compatibility
The ABI version is demoted from 10.0.0 to ~9.0.4~ 9.4.0

Addresses #90

mathias-lang-sociomantic commented:
Shouldn't this target v5.4.x ?
Commit title has a typo.
You might also want to update the documentation (or we can do that later).
Why `.0.4` ?

andreas-bok-sociomantic commented:
> You might also want to update the documentation (or we can do that later). 

in `configure.ac`?

> Why .0.4 ?

Next in line given the previous minor release  `v5.3.4` which had ABI version `9.3.0`

mathias-lang-sociomantic commented:
> in configure.ac?

Yes.

> Next in line given the previous minor release v5.3.4 which had ABI version 9.3.0

The previous one was `9.0.0`. I think it's fine to match the minor release directly, but then you meant `9.4.0`, not `9.0.4` right ?

andreas-bok-sociomantic commented:
> The previous one was 9.0.0. I think it's fine to match the minor release directly, but then you meant 9.4.0, not 9.0.4 right ?

Well, latest release was on v5.3.x (`v5.3.4`), with version in configure.ac = `9.3.0`. This generated `libdrizzle-redux.so.9.0.3`. 
With version in configure.ac = `9.4.0` generates `libdrizzle-redux.so.9.0.4`. 
version in configure.ac = `9.0.1` generates `libdrizzle-redux.so.8.1.0`. Pretty random in my understanding.

andreas-bok-sociomantic commented:
Fixed typos

ben-palmer-sociomantic commented:
Is this the right thing to do here? It seems to me that according to the versioning scheme used by the library that binaries are _not_ compatible.

mathias-lang-sociomantic commented:
![retarded libdrizzle](https://cloud.githubusercontent.com/assets/11289420/21232989/de3f8ff0-c2ed-11e6-8f7b-b51378b5debf.png)

andreas-bok-sociomantic commented:
> Is this the right thing to do here? It seems to me that according to the versioning scheme used by the library that binaries are not compatible.

Yeah, it is a bit weird. The reason is less about standards and more about pragmatics. Apps linked with `so.9` would have to be recompiled with `so.10` and redeployed. 

> @mathias-lang-sociomantic why not make a symlink from `libdrizzle-redux.so.9` to `libdrizzle-redux 10.0.0`. 

Less invasive wrt. to v5.4.0. And introduce a more sane versioning with the next major release.

mathias-lang-sociomantic commented:
The symlink will work too for us. But we should really work that out in the future. Can it be done by the package ?

andreas-bok-sociomantic commented:
> Can it be done by the package ?

Yes. **fpm** provides an arg: 

`--after-install FILE A script to be run after package installation`

Will try it test that it works

andreas-bok-sociomantic committed:
Fix binary compatibility issue

Avoid breaking binary compatibility with v5.4.0
A symlink is created after installation from
/usr/lib/libdrizzle-redux.so.9 to
/usr/lib/libdrizzle-redux.so.10

This is a temporary fix until more a sane ABI
versioning scheme  is implemented

Addresses #90

andreas-bok-sociomantic commented:
Tested with binaries linking against so.9 and so.10 respectively. Works

andreas-bok-sociomantic commented:
Ok feel free to merge

gautam-kotian-sociomantic commented on deb/after_install.sh:
```diff
@@ -0,0 +1,7 @@
+## !/usr/bin/bash
```
That's not really a shebang line, is it?

andreas-bok-sociomantic commented:
fixed dubious shebang and changed `##` to `#` in `after_install.sh`

```diff
diff --git a/deb/after_install.sh b/deb/after_install.sh
index a4cffdd..e467911 100644
--- a/deb/after_install.sh
+++ b/deb/after_install.sh
@@ -1,7 +1,6 @@
-## !/usr/bin/bash
-##
-## Script which is run after install to ensure binary compatibility between
-## this and the previous version of libdrizzle-redux
-##
+# !/usr/bin/env bash
+#
+# Script which is run after install to ensure binary compatibility between
+# this and the previous version of libdrizzle-redux
```

gautam-kotian-sociomantic commented on deb/after_install.sh:
```diff
@@ -0,0 +1,6 @@
+# !/usr/bin/env bash
```
Does it work if there's a space between `#` & `!`?

gautam-kotian-sociomantic replied:
Just confirmed. It doesn't.

EDIT: I meant that this script may work (it's named with a `.sh` extension after all). But in general, a shebang line containing a space between `#` & `!` fails to be interpreted correctly.

gautam-kotian-sociomantic replied:
I'd suggest just using `#!/bin/bash`

gautam-kotian-sociomantic commented on deb/after_install.sh:
```diff
@@ -0,0 +1,6 @@
+# Script which is run after install to ensure binary compatibility between
+# this and the previous version of libdrizzle-redux
+
+sudo ln -sf /usr/lib/libdrizzle-redux.so.10  /usr/lib/libdrizzle-redux.so.9
```
There are two spaces between the last two arguments.

andreas-bok-sociomantic commented:
Updated based on comments from @gautam-kotian-sociomantic 
```diff
diff --git a/deb/after_install.sh b/deb/after_install.sh
index e467911..0edda68 100644
--- a/deb/after_install.sh
+++ b/deb/after_install.sh
@@ -1,6 +1,6 @@
-# !/usr/bin/env bash
+#!/bin/bash
 #
 # Script which is run after install to ensure binary compatibility between
 # this and the previous version of libdrizzle-redux
 
-sudo ln -sf /usr/lib/libdrizzle-redux.so.10  /usr/lib/libdrizzle-redux.so.9
+sudo ln -sf /usr/lib/libdrizzle-redux.so.10 /usr/lib/libdrizzle-redux.so.9
```

gautam-kotian-sociomantic commented:
LGTM.
